web-api-db-ctrl/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+        format!(
+            r##"ALTER SCHEMA public RENAME TO {tenant};
+                -- revoke privileges from public
+                REVOKE ALL ON SCHEMA {tenant} FROM PUBLIC;
+                -- probably unneeded but make sure it's owned by the admin user
+                ALTER SCHEMA {tenant} OWNER TO CURRENT_USER;
+                -- create a new public schema, wo do not grant rights to PUBLIC
+                CREATE SCHEMA public;"##
+        )
+    } else {
+        format!(
+            r##"-- do not use the AUTHORIZATION option, the tenant uses that schema but
+                -- doesn't own it (tenants only own their data, not the structure it's stored in)
+                CREATE SCHEMA IF NOT EXISTS {tenant};"##
+        )
+    };
+
+    tx.execute_many(query.as_str())
+        .try_for_each(|_| future::ready(Ok(())))
+        .await?;
+
+    let query = format!(
+        r##"-- tenant is only allowed to use the schema, they don't own it
+            GRANT USAGE ON SCHEMA {tenant} TO {tenant};
+
+            -- make sure all object we create can be used by tenant
+            -- Note:
+            --   This sets the default privileges for objects created by the user running this
+            --   command, this will not effect the privileges of objects created by other users.
+            ALTER DEFAULT PRIVILEGES IN SCHEMA {tenant}
+                GRANT SELECT, INSERT, UPDATE, DELETE
+                ON TABLES
+                TO {tenant};
+
+            ALTER DEFAULT PRIVILEGES IN SCHEMA {tenant}
+                GRANT USAGE
+                ON SEQUENCES
+                TO {tenant};
+
+            ALTER DEFAULT PRIVILEGES IN SCHEMA {tenant}
+                GRANT EXECUTE
+                ON ROUTINES
+                TO {tenant};
+
+            ALTER DEFAULT PRIVILEGES IN SCHEMA {tenant}
+                GRANT USAGE
+                ON TYPES
+                TO {tenant};
+        "##
+    );
+
+    tx.execute_many(query.as_str())
+        .try_for_each(|_| future::ready(Ok(())))
+        .await?;
+
+    sqlx::query("INSERT INTO management.tenant (tenant_id, is_legacy_tenant) VALUES ($1, $2);")
+        .bind(tenant_id)
+        .bind(is_legacy_tenant)
+        .execute(tx)
+        .await?;
+
+    Ok(())
+}
+
+/// Creates a DB role if it doesn't exist.
+#[instrument(err)]
+async fn create_role_if_not_exists(
+    tx: &mut Transaction<'_, Postgres>,
+    role: &QuotedIdentifier,
+) -> Result<bool, Error> {
+    let mut count = 10;
+    loop {
+        return if !does_role_exist(tx, role).await? {
+            if let Err(err) = create_role(tx, role).await {
+                count -= 1;
+                if err.to_string().contains("tuple concurrently updated") && count > 0 {
+                    continue;
+                } else {
+                    return Err(err);
+                }
+            }
+            Ok(true)
+        } else {
+            Ok(false)
+        };
```

Comment:
This loop/retry logic wasn't triggered a single time when I ran the test multiple times.

I'm not completely sure if it is because it can't fail due to how postgres works or if it's just very unlikely,
so just to be safe I keep this in.

Also this function only is used by the `admin_as_mt_user_hack` so once we remove that we can also remove this function.

Reply:
I ran into this now :tada: 

But it doesn't work :disappointed: .

I forgot that one an sql statement runs into an error in a transaction it will make any further statements in the transaction fail.

So this needs to be done outside of the transaction :thinking: 

But user creation and similar are still transaction in postgres, which we would lose here, which I guess is fine given for what we use it but means it a not fully general function.

Reply:
fixed :rocket: